server/service/appconfig.go

```diff
@@ -218,23 +242,78 @@ func (svc *Service) ModifyAppConfig(ctx context.Context, p []byte) (*fleet.AppCo
 	}
 
 	// if Jira integration settings are modified or added, test it.
-	newJiraSettings := appConfig.Integrations.Jira
-	for i, settings := range newJiraSettings {
-		if i < len(oldJiraSettings) && *oldJiraSettings[i] == *settings {
-			// unchanged
-			continue
+	var newJiraConfig []*fleet.JiraIntegration
+	newJiraByProjectKey := make(map[string]*fleet.JiraIntegration)
+	for i, new := range newAppConfig.Integrations.Jira {
+		// first check for project key uniqueness
+		if _, ok := newJiraByProjectKey[new.ProjectKey]; ok {
+			return nil, ctxerr.Wrap(ctx, &badRequestError{message: fmt.Sprintf("duplicate Jira integration for project key %s", new.ProjectKey)})
+		}
+		newJiraByProjectKey[new.ProjectKey] = new
+
+		// check if existing integration is being edited
+		if old, ok := storedJiraByProjectKey[new.ProjectKey]; ok {
+			if old == *new {
+				newJiraConfig = append(newJiraConfig, &old)
+				// no further validtion for unchanged integration
+				continue
+			}
+			// use stored API token if request does not contain new token
+			if new.APIToken == "" || new.APIToken == "********" {
+				new.APIToken = old.APIToken
+			}
 		}
 
 		// new or updated, test it
-		if err := svc.makeTestJiraRequest(ctx, settings); err != nil {
+		if err := svc.makeTestJiraRequest(ctx, new); err != nil {
 			return nil, ctxerr.Wrapf(ctx, err, "Jira integration at index %d", i)
 		}
+		newJiraConfig = append(newJiraConfig, new)
 	}
+	appConfig.Integrations.Jira = newJiraConfig
+
+	// if Zendesk integration settings are modified or added, test it.
+	var newZendeskConfig []*fleet.ZendeskIntegration
+	newZendeskByGroupID := make(map[int64]*fleet.ZendeskIntegration)
+	for i, new := range newAppConfig.Integrations.Zendesk {
+		// first check for group id uniqueness
+		if _, ok := newZendeskByGroupID[new.GroupID]; ok {
+			return nil, ctxerr.Wrap(ctx, &badRequestError{message: fmt.Sprintf("duplicate Zendesk integration for group id %v", new.GroupID)})
+		}
+		newZendeskByGroupID[new.GroupID] = new
+
+		// check if existing integration is being edited
+		if old, ok := storedZendeskByGroupID[new.GroupID]; ok {
+			if old == *new {
+				newZendeskConfig = append(newZendeskConfig, &old)
+				// no further validtion for unchanged integration
+				continue
+			}
+			// use stored API token if request does not contain new token
+			if new.APIToken == "" || new.APIToken == "********" {
```

Comment:
I believe the check makes sense given current protocol/API, but this check does not look good (mostly error-prone).

Short term solution I can think of (not saying we should do it on this PR):
- Have `APIToken be *string `json:"api_token,omitempty"``? (and let the UI display `"********"` or whatever and not set the field in the `PATCH` request when it's not set.)

Long term solution I can think of:
- Separate endpoint for integrations configuration? (ModifyAppConfig is 153 lines long and will continue to grow with every integration.)

Reply:
Agreed. This is intended to be a very short-term accommodation for the frontend and will be revisited as we move to a dedicated endpoint to manage integrations.  I'll add a comment to make this explicit for future reference.